src/Microsoft.AspNet.Mvc.Core/ActionConstraints/ConsumesAttribute.cs

```diff
@@ -0,0 +1,153 @@
+            if (requestContentType == null)
+            {
+                var isActionWithoutConsumeConstraintPresent = context.Candidates.Any(
+                    candidate => candidate.Constraints == null ||
+                    !candidate.Constraints.Any(constraint => constraint is IConsumesActionConstraint));
+
+                return !isActionWithoutConsumeConstraintPresent;
+            }
+
+            if (ContentTypes.Any(c => c.IsSubsetOf(requestContentType)))
+            {
+                return true;
+            }
+
+            var firstCandidate = context.Candidates[0];
+            if (firstCandidate != context.CurrentCandidate)
+            {
+                // If the current candidate is not same as the first candidate,
+                // we need not probe other candidates to see if they apply.
+                // Only the first candidate is allowed to probe other candidates and based on the result select itself.
+                return false;
+            }
+
+            // Run the matching logic for all IConsumesActionConstraints we can find, and see what matches.
+            // 1). If we have a unique best match, then only that constraint should return true.
+            // 2). If we have multiple matches, then all constraints that match will return true
+            // , resulting in ambiguity(maybe).
+            // 3). If we have no matches, then we choose the first constraint to return true.It will later return a 415
+            foreach (var candidate in context.Candidates)
+            {
+                if (candidate == firstCandidate)
+                {
+                    continue;
+                }
+
+                var tempContext = new ActionConstraintContext()
+                {
+                    Candidates = context.Candidates,
+                    RouteContext = context.RouteContext,
+                    CurrentCandidate = candidate
+                };
+
+                if (candidate.Constraints == null || candidate.Constraints.Count() == 0 ||
+                    candidate.Constraints.Any(constraint => constraint is IConsumesActionConstraint &&
+                                                            constraint.Accept(tempContext)))
+                {
+                    // There is someone later in the chain which can handle the request.
+                    // end the process here.
+                    return false;
+                }
+            }
+
+            // There is no one later in the chain that can handle this content type return a false positive so that
+            // later we can detect and return a 415.
+            return true;
+        }
+
+        private List<MediaTypeHeaderValue> GetContentTypes(string firstArg, string[] args)
+        {
+            var contentTypes = new List<MediaTypeHeaderValue>();
+            contentTypes.Add(MediaTypeHeaderValue.Parse(firstArg));
+            foreach (var item in args)
```

Comment:
not fixing the select ... it would throw for invalid header type, that sounds like a test for MediaTypeHeaderValue and not consumes, however i added a test to cover that.